### ISSUE TITLE
MSSQL: The ORDER BY clause is invalid in subqueries

### ISSUE THREAD

**Issue by YuxiLing** [PRQL/prql#1854]
Hi, I'm using PRQL to generate MSSQL queries. But I'm facing the following issue.

PRQL version: 0.5.0

Here is an example:

prql code:
```
prql target:sql.mssql
from tb1
join side:inner it=tb2 [tb1.c1 == it.c2]
sort [tb1.c1]
aggregate [d0 = sum tb1.c13,d1 = max tb1.c16]
```

mssql code:
```
WITH table_1 AS (
  SELECT
    SUM(tb1.c13) AS d0,
    MAX(tb1.c16) AS d1,
    tb1.c1
  FROM
    tb1
    JOIN tb2 AS it ON tb1.c1 = it.c2
  ORDER BY
    tb1.c1
)
SELECT
  d0,
  d1
FROM
  table_1

-- Generated by PRQL compiler version:0.5.0 (https://prql-lang.org)
```

Error reported by MS SQL Server:
```
Failed to execute query. Error: The ORDER BY clause is invalid in views, inline functions, derived tables, subqueries, and common table expressions, unless TOP, OFFSET or FOR XML is also specified.
```

I'm thinking that there might be some issues in generating mssql code inside PRQL compiler.
Thank you for your clarification in advance.

**Comment by max-sixty** [comment]
Thanks for the excellent bug report @YuxiLing .

For context, we're currently at the stage where there will be bug like this, particularly in some of our less commonly used dialects. These reports help us improve.

For this one, while it's MSSQL-specific, we could consider not having `ORDER BY` unless it's on the outer-most query. Related to https://github.com/PRQL/prql/issues/1363.

FWIW moving the `sort` to the final line does work

**Comment by YuxiLing** [comment]
Thank you for your reply.

For this issue, the reason I'm thinking is that MSSQL requires `ORDER BY` in a view to being with some further operations, like `TOP`, `OFFSET`, which are equal to the `take` clause in prql. 

This is an immature idea from my side. Maybe we can move the `sort` to the final line when there is no `take` operation nearby. If there is no `take` clause, the position of `sort` doesn't affect the result.

Actually, I am trying to help you make PRQL more robust by finding potential issues. Hope this would be helpful.

**Comment by max-sixty** [comment]
> Actually, I am trying to help you make PRQL more robust by finding potential issues. Hope this would be helpful.

Very much so, this is helpful, and please continue putting issues like this in!

**Comment by aljazerzen** [comment]
Yup, we should never produce ORDER BY for SQL subqueries, since they are never guaranteed to have any effect. MSSQL is just more terse here and actually produces and error where other engines any be returning unexpected results.

But we should infer that the final PRQL relation has sorting and apply ORDER BY to the main query, as discussed in #1363.

**Comment by max-sixty** [comment]
@aljazerzen I've been thinking about issues which would require less context than others, as discussed earlier today.

I wonder whether this could be an issue where someone would _only_ need to understand RQ to work on it — i.e. given some RQ with `sort`, we can decide where the `ORDER BY` goes. There'd be no need to understand PL or parsing or anything on that side of the compiler.

Lmk if you have thoughts. Maybe I'll take it as an example if you think that's correct.

**Comment by aljazerzen** [comment]
That's true, I'll open a good-first-issue

**Comment by aljazerzen** [comment]
Thinking about it, it is not straightforward enough that I could just write down what to do.

So instead, I'll post a few examples what should happen here:

```elm
let a = (from x | sort y)

from a
```
```sql
WITH a AS (SELECT * FROM x)
SELECT * FROM a ORDER BY y
-- ORDER BY must be in second SELECT only
```

----

```elm
from a
sort b   # initial sort
take 10  # this forces the pipeline to be split
join c []
```
```sql
WITH table_1 AS (SELECT * FROM a LIMIT 10)
SELECT table_1.*, c.* AS x FROM table_1 JOIN c ON TRUE
ORDER BY b
-- ORDER BY must be in second SELECT only
```

----

```elm
from a
sort b
take 10    # this forces the pipeline to be split
group x (aggregate count)  # group resets sort order
```
```sql
WITH table_1 AS (SELECT * FROM a LIMIT 10)
SELECT COUNT(*) FROM table_1 GROUP BY x
```

**Comment by max-sixty** [comment]
I was thinking — as part of trying to modularize things — that we could show just the changes to RQ. Then someone doesn't need to understand the full pipeline — instead just a function over the RQ would be sufficient.

For example, take the first query; IIUC this would require this diff:

```diff
diff --git a/x.yaml b/x.yaml
index 7f8eff8..3b2c9ae 100644
--- a/x.yaml
+++ b/x.yaml
@@ -1,52 +1,52 @@
 def:
   version: null
   other: {}
 tables:
   - id: 0
     name: null
     relation:
       kind:
         ExternRef:
           LocalTable: x
       columns:
         - Single: y
         - Wildcard
   - id: 1
     name: a
     relation:
       kind:
         Pipeline:
           - From:
               source: 0
               columns:
                 - - Single: y
                   - 0
                 - - Wildcard
                   - 1
               name: x
-          - Sort:
-              - direction: Asc
-                column: 0
           - Select:
               - 0
               - 1
       columns:
         - Single: y
         - Wildcard
 relation:
   kind:
     Pipeline:
       - From:
           source: 1
           columns:
             - - Single: y
               - 2
             - - Wildcard
               - 3
           name: a
       - Select:
           - 2
           - 3
+      - Sort:
+          - direction: Asc
+            column: 0
   columns:
     - Single: y
     - Wildcard
```

...the others would require similar diffs (and someone would need to write the actual logic for deciding, but my assertion is that that part is actually not the most difficult piece).

---

Clearly this is more verbose than the examples with the PRQL & SQL. But it's arguably much more contained & modularized....

**Comment by aljazerzen** [comment]
I believe this was closed with #1363

This query:
```elm
from tb1
join side:inner it=tb2 [tb1.c1 == it.c2]
sort [tb1.c1]
aggregate [d0 = sum tb1.c13,d1 = max tb1.c16]
```

Now produces:
```sql
WITH table_3 AS (
  SELECT
    SUM(tb1.c13) AS d0,
    MAX(tb1.c16) AS d1,
    tb1.c1
  FROM
    tb1
    JOIN tb2 AS it ON tb1.c1 = it.c2
),
table_2 AS (
  SELECT
    d0,
    d1
  FROM
    table_3 AS table_1
)
SELECT
  d0,
  d1
FROM
  table_2 AS table_0

-- Generated by PRQL compiler version:0.8.1 (https://prql-lang.org)
```
... which does contain an-unnecessary CTE, but works as expeceted.

**Comment by max-sixty** [comment]
Awesome!
